routing/result_interpretation.go

```diff
@@ -394,6 +394,48 @@ func (i *interpretedResult) processPaymentOutcomeIntermediate(
 	case *lnwire.FailExpiryTooSoon:
 		reportAll()
 
+	// If FailInvalidBlinding is received, then there has been a failure
+	// within a blinded route. The introduction node is always responsible
+	// for reporting errors within a blinded route (to protect the privacy
+	// of members of the route), so it will always be the reporting node.
+	// We penalize all of the nodes _after_ the introduction point, so that
+	// the blinded route is penalized.
+	//
+	// The introduction node has no incentive to falsely report errors to
+	// sabotage the blinded route because:
+	//   1. Its ability to route this payment is strictly tied to the
+	//      blinded route.
+	//   2. The pubkeys in the blinded route are ephemeral, so doing so
+	//      will have no impact on the nodes beyond the individual payment.
+	//
+	// If a node that was *not* the introduction node in a blinded route
+	// reports this error code, we penalize the node because it's not
+	// supposed to be sending this error.
+	case *lnwire.FailInvalidBlinding:
+		// Penalize the failing node if it was not the introduction
+		// point of a blinded route.
+		if route.Hops[errorSourceIdx-1].BlindingPoint == nil {
+			// If there are any hops before the erring node, assign
+			// success to their pairs. We don't want to assign
+			// success to the incoming edge for the failing node,
+			// because it could have sent a blinding point to force
+			// this error (and we're about to penalize it).
+			if errorSourceIdx >= 2 {
+				i.successPairRange(route, 0, errorSourceIdx-1)
+			}
+
+			i.failNode(route, errorSourceIdx)
+
+			return
+		}
+
+		// Everything up until the error source forwarded the payment
+		// correctly.
+		i.successPairRange(route, 0, errorSourceIdx-1)
+		i.failPairRange(
```

Comment:
> I feel like we should use failPairBalance only for the last hop, as this will already invalidate the blinded part of the route for the amount that was tried

Oh yeah, this is much nicer. WDYT we should do in the case of a single hop blinded route (`Introduction -> Blinded Recipient`), because `failPairBalance` is only effective on intermediate nodes rn? afaik we don't have a way to penalize the blinded recipient without penalizing the intro node, so just set a final failure so that we give up? 

> Otherwise this may be gameable by providing faulty invoices that would worsen/improve introduction nodes' routing reputation.

I do wonder how much we _actually_ need to worry about this? To "poison" an introduction node, the malicious receiver needs to get the sender to pay that invoice. Which would only happen if the sender wants to make that payment (buy something / send money etc), and then surely the receiver is more incentivized to "shut up and take my money"? Not disagreeing with leaving the intro node untouched (makes sense for node score), just not sure how much of a threat it would be.

Thanks for taking a look!

Reply:
> Oh yeah, this is much nicer. WDYT we should do in the case of a single hop blinded route (Introduction -> Blinded Recipient), because failPairBalance is only effective on intermediate nodes rn? afaik we don't have a way to penalize the blinded recipient without penalizing the intro node, so just set a final failure so that we give up?

Yes, that one is tricky, but it seems reasonable to fail the payment, since the receiver should have added a route hint that has liquidity. Another case which is not handled right now is the case where we see `FailInvalidBlinding` from a node after the introduction point (spec violated twice?). In that case we probably would like to punish the introduction node for not letting the error originate from it (thereby also invalidating the blinded route). Since all of this is really complicated I tried check any edge cases as well (https://github.com/bitromortac/lnd/tree/re-8095-231018-11-blinded-error).

Reply:
> Yes, that one is tricky, but it seems reasonable to fail the payment, since the receiver should have added a route hint that has liquidity. 

I think this might get a little tricky with multi-blinded path support, but I think this is a good start for the single-path case.

Thanks for going through all the edge cases! Updated code to account for single hop case + errors after the introduction node. Also updated PR description with overview of each case - lmk if anything is missing / different to your interpretation.